radix-engine/src/blueprints/intent_hash_store/package.rs

```diff
@@ -0,0 +1,280 @@
+                        virtual_lazy_load_functions: btreemap!(),
+                        functions,
+                    },
+                },
+
+                royalty_config: RoyaltyConfig::default(),
+                auth_config: AuthConfig {
+                    function_auth: btreemap!(
+                    ),
+                    method_auth: MethodAuthTemplate::Static {
+                        auth: btreemap!(),
+                        outer_auth: btreemap!(),
+                    },
+                },
+            }
+        );
+
+        PackageDefinition { blueprints }
+    }
+
+    #[trace_resources(log=export_name)]
+    pub fn invoke_export<Y>(
+        export_name: &str,
+        input: &IndexedScryptoValue,
+        api: &mut Y,
+    ) -> Result<IndexedScryptoValue, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        match export_name {
+            INTENT_HASH_STORE_CREATE_EXPORT_NAME => {
+                api.consume_cost_units(FIXED_LOW_FEE, ClientCostingReason::RunNative)?;
+
+                let input: IntentHashStoreCreateInput = input.as_typed().map_err(|e| {
+                    RuntimeError::ApplicationError(ApplicationError::InputDecodeError(e))
+                })?;
+
+                let rtn = IntentHashStoreBlueprint::create(input.address_reservation, api)?;
+
+                Ok(IndexedScryptoValue::from_typed(&rtn))
+            }
+
+            _ => Err(RuntimeError::ApplicationError(
+                ApplicationError::ExportDoesNotExist(export_name.to_string()),
+            )),
+        }
+    }
+}
+
+#[derive(Debug, Clone, ScryptoSbor)]
+pub enum IntentHashStatus {
+    CommittedSuccess,
+    CommittedFailure,
+    Cancelled,
+}
+
+#[derive(Debug, Clone, ScryptoSbor)]
+pub struct IntentHashStoreSubstate {
+    pub start_epoch: u64,
+    pub start_partition: u8,
+
+    // parameters
+    pub partition_range_start: u8,
+    pub partition_range_end: u8,
+    pub epochs_per_partition: u64,
+}
+
+impl IntentHashStoreSubstate {
+    pub fn partition_of(&self, epoch: u64) -> Option<u8> {
+        // Check if epoch is within range
+        let num_partitions = self.partition_range_end - self.partition_range_start + 1;
+        let max_epoch_exclusive =
+            self.start_epoch + num_partitions as u64 * self.epochs_per_partition;
+        if epoch < self.start_epoch || epoch >= max_epoch_exclusive {
+            return None;
+        }
+
+        // Calculate the destination partition number
+        let mut partition_number =
+            self.start_partition as u64 + (epoch - self.start_epoch) / self.epochs_per_partition;
+        if partition_number > self.partition_range_end as u64 {
+            partition_number -= num_partitions as u64;
+        }
+
+        assert!(partition_number >= self.partition_range_start as u64);
+        assert!(partition_number <= self.partition_range_end as u64);
+
+        Some(partition_number as u8)
+    }
+
+    pub fn advance(&mut self) -> u8 {
```

Comment:
This API is a bit confusing I think. To use it correctly leaks a lot of implementation details into (eg) the transaction executor.

I'd rather use a `handle_new_epoch(epoch: Epoch) -> EpochChangeHandling` which decides if it wants to advance this? With:
```
pub enum EpochChangeOutcome {
    None,
    DeleteFullyExpiredPartition(u8),
}
```

Or if we'd rather keep this, we should add a Rust doc explaining how to use this safely.

Reply:
Haha, will just add comments. 

The motivation here is primarily to follow [ring buffer design](https://en.wikipedia.org/wiki/Circular_buffer), and not coupled with epoch update handling.

Transaction executor is already full of logic lol.

Reply:
Hmm but I'd rather the deletion logic was better defined by the protocol and didn't depend on the transaction validator logic, or require loops in the clear up process.